### PR TITLE
feat: strict mode for warnings

### PR DESCRIPTION
**Pull request by varl**
Allow d2-style to print warnings from ESLint and introduce a concept of
strict and non-strict modes.

Strict mode: Set a zero tolerance on warnings by making ESLint exit with
a non-zero code if there are any warnings. Warnings will be shown in
strict mode.

Non-strict mode: Allow ESLint to exit with code zero when there are
warnings. Warnings are essentially ignored in this mode which matches
the behavior today.

This allows gradual introduction of new rules, that may later become
errors.

In strict mode a developer is expected to either:

a)  fix the warning
b)  explicitly mark the code that triggered the warning with an
eslint-disable comment.

A good disable directive uses the comment syntax for ESLint:

    /* eslint-disable [rule] -- [comment] */

E.g.

    /* eslint-disable-next-line no-console -- we need to log here */

When rules graduate from "warning" to "error" the code base should be
ready for the change.

### CODE REVIEW

**Comment by varl** [comment]
Supersedes #366.

This is a middle-ground that:

- Is not a breaking change, the default functionality is the same as today: no warnings are shown on e.g. `d2-style js check`.
- Allows warnings to be displayed in strict mode: `d2-style js check --strict`.
- Strict mode will exit with a non-zero code if there are warnings.
- We can add rules as warnings, and they don't impact users until they are upgraded to errors.
- We can run strict mode on our apps after adding new rules to see how much effort it is going to take if we upgrade a rule to errors and plan accordingly.

File: docs/faq.md [hunk 1]
```diff
@@ -14,3 +14,31 @@ You can reinstall it using 'npm install husky --save-dev' or delete this hook
 ```
 
 Ensure that your node and git versions satisfy the above requirements. You can check this by running `git --version` and `node --version` from your terminal.
+
+# What does an Error and Warnings mean?
```
Comment (varl) [inline docs/faq.md:18]:
```suggestion
# What does Error and Warning mean?
```

**Comment by HendrikThePendric** [comment]
My perspective is this:
- I was happy with how `cli-style` worked, I didn't really see any need for structural changes
- I would be very much in favour of introducing plugins that help us catch problems with `i18n` and `a11y`
- But I expect apps and libs are going to have to make **a lot of changes** when they upgrade to the version of cli-style containing these rules. I think to the extend that it could probably prevent apps from upgrading.
- So I think we need to offer a gentle upgrade path somehow. Showing warnings instead of errors would be one way to achieve this. But we could also choose a completely different route, for example:
    - Custom commands to lint the codebase for `a11y` or `i18n`
    - Possibility to switch the `i18n` and `a11y` rules on and of (obviously this is possible already, but we could offer some convenience methods).

I don't really have an opinion about the broader discussion about warnings VS errors, so will refrain from commenting on that.

**Comment by varl** [comment]
>     * I was happy with how `cli-style` worked, I didn't really see any need for structural changes

Me too, so this PR doesn't change anything in terms of usage.

> I feel like with this PR we're trying to implement a generic solution for a specific problem.

This PR only provides a way to see rules configured as warnings in the console. Today they are completely ignored with no way of seeing them at all. Maybe `strict` is a poor choice of name? 

> I fear to the extend it could probably prevent apps/libs from upgrading the package.

I share this fear, and would like to deal with it elsewhere, so this PR keeps things as they are.

> Showing warnings instead of errors would be one way to achieve this. But we could also choose a completely different route, for example:

Nice examples of alternatives! Those could indeed be ways to go about it. Internally those options would require significant structural changes in how commands are organized, as well as how we configure and run the toolchains.

**Comment by varl** [comment]
> ```
> --fail-on-warning (implicitly enables the flag below)
> --show-warnings
> ```

I guess these would be more acceptable to @HendrikThePendric as well (no structural/conceptual changes)?

**Comment by ismay** [comment]
This seems similar to an issue that was filed with eslint: https://github.com/eslint/eslint/issues/2309. I think it's relevant to our discussion here.

What I was proposing was basically to let eslint do its thing, without letting warnings fail anything, and without swallowing them. The way I understand the proposed strict mode here (which is similar to the proposed strict mode in the issue), is that it would exit with a non-zero exit code on warnings. As the maintainer in the issue there mentioned I also feel that that logic blurs the line between errors and warnings (a warning by definition shouldn't fail anything, then we should just make it an error).

Basically what I meant is just for us to remove the `--quiet` flag. To me that seems more in line with the eslint ecosystem (and stylelint as well). But I'm not sure if that goes against a requirement that we have, let me know if it does or if I'm misunderstanding something.

**Comment by varl** [comment]
> But I'm not sure if that goes against a requirement that we have, let me know if it does or if I'm misunderstanding something.

The way d2-style was designed was to be a gatekeeper. That was its reason for existing, it has no nuance. Either it is good or it is bad.

What it is evolving to become is a developer support tool. And guidance requires more nuance than enforcing, so I see where the need for both errors and warnings in that sense.

> To me that seems more in line with the eslint ecosystem (and stylelint as well).

The tools that d2-style used to accomplish its agenda were intended to be implementation details that are completely obscured from the user. This was changed when developers requested to be able to see lint/formatting errors in their editors.

> edit: Allowing `--show-warnings` would basically do the same as what I'm talking about. Seems good as well. Though by making it a flag I guess then we'd have to have the discussion where to enable it. I'd just make it the default.

My intent was for the user to use it manually to check if there are any "soon to be problems", not that it should be enabled anywhere.

**Comment by varl** [comment]
I need to finish up my implementation of the `d2-style ci` command.

That would be my enforcer of good and bad. The other commands could be allowed more nuance to decide what is ugly.

**Comment by ismay** [comment]
> > But I'm not sure if that goes against a requirement that we have, let me know if it does or if I'm misunderstanding something.
> 
> The way d2-style was designed was to be a gatekeeper. That was its reason for existing, it has no nuance. Either it is good or it is bad.
> 
> What it is evolving to become is a developer support tool. And guidance requires more nuance than enforcing, so I see where the need for both errors and warnings in that sense.
> 
> > To me that seems more in line with the eslint ecosystem (and stylelint as well).
> 
> The tools that d2-style used to accomplish its agenda were intended to be implementation details that are completely obscured from the user. This was changed when developers requested to be able to see lint/formatting errors in their editors.
> 
> > edit: Allowing `--show-warnings` would basically do the same as what I'm talking about. Seems good as well. Though by making it a flag I guess then we'd have to have the discussion where to enable it. I'd just make it the default.
> 
> My intent was for the user to use it manually to check if there are any "soon to be problems", not that it should be enabled anywhere.

I see what you mean. Yeah it seems like we both had a slightly different perception of the design goals of cli-style, thanks for explaining!
